### PR TITLE
[MAGETWO-54158]:+ Corrected ACL rules for the sendfriend module and s…

### PR DESCRIPTION
**Pull request by jspoe**
### Description

Added an acl.xml to the send-friend module
Added upgrade data script for backward compatibility for ACL-roles that were already selected for users.

Changed the acl.xml for the backend module
Added upgrade data script for backward compatibility for ACL-roles that were already selected for users.

### Fixed Issues (if relevant)

1. magento/magento2#MAGETWO-54158: ACL not defined for Magento_Backend::web

### Manual testing scenarios

1. Install Magento latest develop version
2. Go to Admin side
3. Click "System > User Roles"
4. Click "Add new Role" enter Role name
5. In left side choose Role Resources and check "Dashboard", "Web Section", "Email to a friend", "Currency Setup", "Store E-mail addresses"
6. Save Role
7. Click "System > All Users"
8. Click "Add new User"
9. Enter data in required fields
10. In left side click "User Role" and choose early created role.
11. Click "Save"
12. Sign out and Log in using credentials of new user
13. Click "Stores > Configuration"

Expected result:
All the tabs that were selected in the role should be visible

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=9628) <br/>All committers have signed the CLA.

File: app/code/Magento/Config/Setup/UpgradeData.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        if (version_compare($context->getVersion(), '2.0.1', '<')) {
+            $this->upgradeAcl($setup);
+        }
+    }
+
+    /**
+     * @param ModuleDataSetupInterface $setup
+     * @return void
+     */
+    protected function upgradeAcl(ModuleDataSetupInterface $setup)
```
Comment (ishakhsuvarov) [inline app/code/Magento/Config/Setup/UpgradeData.php]: Is there any reason for this method is protected?  Looks like private is more suitable in this case.

File: app/code/Magento/Config/Setup/UpgradeData.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+use Magento\Framework\Setup\ModuleDataSetupInterface;
+use Magento\Framework\Setup\UpgradeDataInterface;
+
+/**
+ * @codeCoverageIgnore
+ */
+class UpgradeData implements UpgradeDataInterface
+{
+
+    /**
```
Comment (ishakhsuvarov) [inline app/code/Magento/Config/Setup/UpgradeData.php:19]: `{@inheritdoc}` is sufficient in this case.

File: app/code/Magento/Config/Setup/UpgradeData.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+     */
+    public function upgrade(ModuleDataSetupInterface $setup, ModuleContextInterface $context)
+    {
+        if (version_compare($context->getVersion(), '2.0.1', '<')) {
+            $this->upgradeAcl($setup);
+        }
+    }
+
+    /**
+     * @param ModuleDataSetupInterface $setup
```
Comment (ishakhsuvarov) [inline app/code/Magento/Config/Setup/UpgradeData.php]: Would be nice to add a comment which explains background under this action.

File: app/code/Magento/Config/etc/acl.xml [hunk 4]
```diff
@@ -12,16 +12,15 @@
                 <resource id="Magento_Backend::stores">
                     <resource id="Magento_Backend::stores_settings">
                         <resource id="Magento_Config::config" title="Configuration" translate="title" sortOrder="20">
-                            <resource id="Magento_Config::advanced" title="Advanced Section" translate="title" sortOrder="90" />
+                            <resource id="Magento_Backend::advanced" title="Advanced Section" translate="title" sortOrder="90" />
```
Comment (ishakhsuvarov) [inline app/code/Magento/Config/etc/acl.xml]: Did you consider introducing these resources in the `Backend/etc/acl.xml`?

File: app/code/Magento/SendFriend/Setup/UpgradeData.php [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        if (version_compare($context->getVersion(), '2.0.1', '<')) {
+            $this->upgradeAcl($setup);
+        }
+    }
+
+    /**
+     * @param ModuleDataSetupInterface $setup
+     * @return void
+     */
+    protected function upgradeAcl(ModuleDataSetupInterface $setup)
```
Comment (ishakhsuvarov) [inline app/code/Magento/SendFriend/Setup/UpgradeData.php]: Please consider making this method private

**Review by ishakhsuvarov — CHANGES_REQUESTED** [review]

**Comment by ishakhsuvarov** [comment]
Hi @jspoe 
Would you like to continue work on this PR or should someone else pick it up?

**Comment by jspoe** [comment]
Hi @ishakhsuvarov 

Thank you for your feedback!
I will pickup the requested changes in a couple of days. Should not take a lot of time.

**Comment by ishakhsuvarov** [comment]
@jspoe Thanks, we will proceed with review shortly.

**Review by ishakhsuvarov — APPROVED** [review]

**Comment by ishakhsuvarov** [comment]
@jspoe It looks like these changes introduce some unwanted dependencies between the modules in question. We will look into possible solution and let you know how we can fix it.

**Comment by jspoe** [comment]
Yes I understand. It is either part of the Magento_Config module or the Magento_Backend module.
It should not be both. Then there is also the part of the setup script where there is a (small) dependency for to map the roles for backward compatibility.

Please let me know as soon as possible if I can fix this in a better way for you.

**Comment by ishakhsuvarov** [comment]
@jspoe Looks like the best solution in this case would be to move upgrade scripts to the specifically created modules, something like SendfriendAcl or SendfriendAuthorization, which would have correct dependencies and would allow other modules to be disabled/removed safely.
Would you like to continue on that, since it is some more work to do, or do you want someone else to pick up?
Thanks!

**Comment by jspoe** [comment]
Hi @ishakhsuvarov,

I agree, and I will pick up the requests myself if possible.
I will push a change as soon as I can.

Thanks!

**Comment by ishakhsuvarov** [comment]
@jspoe Thank you

**Comment by ishakhsuvarov** [comment]
@jspoe I am closing this PR for now due to inactivity. Please feel free to update and reopen whenever you wish to continue work on it.
Thank you!

**Pull request closed without merging**